src/Data/Traversable.js

```diff
@@ -3,58 +3,32 @@
 // jshint maxparams: 3
 
 exports.traverseArrayImpl = function () {
-  function Cont(fn) {
-    this.fn = fn;
-  }
-
-  var emptyList = {};
-
-  var ConsCell = function (head, tail) {
-    this.head = head;
-    this.tail = tail;
-  };
+  function array1(a) {return [a];}
 
-  function consList(x) {
-    return function (xs) {
-      return new ConsCell(x, xs);
-    };
+  function array2(a) {
+    return function (b) {return [a, b];};
   }
 
-  function listToArray(list) {
-    var arr = [];
-    while (list !== emptyList) {
-      arr.push(list.head);
-      list = list.tail;
-    }
-    return arr;
+  function concat2(xs) {
+    return function (ys) {return xs.concat(ys);};
   }
 
   return function (apply) {
     return function (map) {
       return function (pure) {
         return function (f) {
-          var buildFrom = function (x, ys) {
-            return apply(map(consList)(f(x)))(ys);
-          };
-
-          var go = function (acc, currentLen, xs) {
-            if (currentLen === 0) {
-              return acc;
-            } else {
-              var last = xs[currentLen - 1];
-              return new Cont(function () {
-                return go(buildFrom(last, acc), currentLen - 1, xs);
-              });
-            }
-          };
-
           return function (array) {
-            var result = go(pure(emptyList), array.length, array);
-            while (result instanceof Cont) {
-              result = result.fn();
+            function go(bot, top) {
+              switch (top - bot) {
+              case 0: return pure([]);
+              case 1: return map(array1)(f(array[bot]));
+              case 2: return apply(map(array2)(f(array[bot])))(f(array[bot + 1]));
+              default:
+                var pivot = Math.floor((top + bot) / 2);
```

Comment:
By the way, please feel free to take these commits and put them back into your PR if you want, and then I could close this.